### ISSUE TITLE
Pipeline evacuation optimizations.

### ISSUE THREAD

**Issue by antialize** [thomasmoelhave/tpie#208]
Before commit commit 8d09b916a09d0ebe4780baa0142b59a6104b6aa1, pipeline evacuation was completely broken. With this new commit it sort of works, but there are some small issues. 

We introduced an dependency edge saying that a note in one phase shares memory with a note in another phase, now if these two phases are not executed constitutively, the shared memory will have to be evacuated to disk, since some other phase running between the two phases could need the memory. Obviously we want to minimize the number of evacuations, but how?

Let a normal dependency between two phase be represented by a _black_ edge, and let a memory sharing dependency be represented by a _red_ edge if the memory can be evacuated and _green_ if it cannot be evacuated.  We say that a none _black_ edge is satisfied if its two end points are consecutive in the topological order, so the objective is to maximize the number of satisfied edges. Also we must satisfy ALL _green_ edges, if this is not possible the input is mall formed, and someone has to implement an evacuate method somewhere.

First note that a non _black_ edges cannot be satisfied if there exists an alternative path from its source to its destination (with length at least 2), so any such _red_ edge can be recolored to black, if there is any such _green_ edge the input is invalid.
 
Next note that for any node we can satisfy at most one outgoing edge, and at most one incoming edge.  So lets construct a bipartite graph, with all notes with outgoing none _black_ edges on the left, and all nodes with incoming none _black_ edges on the right. Now construct a bipartite maximal matching (where all _green_ edges are pre matched, if it is not possible to pre match all the _green_ edges then the input is invalid). Now the claim is that it is possible to construct a topological order where all nodes of matched edges occur consecutively and that this order satisfies the maximal number of edges.

The top order can be constructed, by constructing a top order of a version of the graph where all the matched edges have been contracted.

**Comment by Tyilo** [comment]
Given a simple diamond dependency phase graph where all edges are red:

```
1 -> 2
 \    \
  v    v
  3 -> 4
```

and the found matching is `{1, 2}, {2, 4}` how would you construct the top order?

Wouldn't the contraction give you a cyclical graph:
```
1,2,4 <-> 3
```

**Comment by Tyilo** [comment]
Also see https://github.com/thomasmoelhave/tpie/tree/optimize_evacuation

**Comment by antialize** [comment]
I see the example is a counter example to the proposed solution. It appears that for a given path from some a to some b, it is not possible to satisfy the whole path if there exists an alternative path from a to b.  I have not been able to come up with a good solution to the problem, perhaps it is NP hard (I have tried to make a reduction from 3-sat, but have not been able to come up with good gadgets). Perhaps we should find a whitebord and find a solution?

**Comment by Tyilo** [comment]
Red edges are not handled yet, but with my latest commit we will satisfy all green edges and throw an exception if that is not possible.

**Comment by Tyilo** [comment]
It is not clear how to handle phases with flush priorities. Should we prioritizing satisfying red edges over satisfying flush priorities?

**Comment by antialize** [comment]
I dont think anyone has ever set a flush priority (it seems too hard). If we want to go with a gready approach and keep flush priorities, then it is quite easy. But if we (as we should) use a better strategy then we should probably drop the priorities, or maximize the sum of satisfied priorities.

**Comment by Tyilo** [comment]
>I dont think anyone has ever set a flush priority (it seems too hard).]

Do you mean it isn't used? @Mortal talked about it being used in Terrastream?

>If we want to go with a gready approach and keep flush priorities, then it is quite easy. But if we (as we should) use a better strategy then we should probably drop the priorities, or maximize the sum of satisfied priorities.

I have implemented a better algorithm in [optimize_evacuation](https://github.com/thomasmoelhave/tpie/tree/optimize_evacuation), which just ignores flush priorities.

Do you think the following solution would be good enough:

- Satisfy green edges and contract them
- Satisfy as many red edges as possible with green edges contracted and contract them
- Use flush priorities to get a top order for the contracted graph
- Un-contract the contracted red edges
- Un-contract the green edges

**Comment by antialize** [comment]
In terrastream i get:
```bash
>>> git grep set_flush_priority | wc -l
0
```

So it is safe to say that no one uses priorities, i think it is too difficult to maintain in practice. If we have two concurent sorters such that one sorts a small amount of data and one sort a large amount, we will give them different memory priorities, we have not thought to also give them different flush priorities.  Perhaps it would be best to not have a flush priority and instead use the memory priority also as a flush priority.

I think your solution is good enough, but you could also implement an optimal solution when then number of red edges is below some small number (say 20) as is often the case. 

Just do something like:
```python
bestScore = 0
bestG = None 
for i in range(2**n):
   good = False
   Gp = G
   score = 0
   for j in range(n):
      if good and (i & (1 << j)):
        good = Gp.contract(j)
        score += G.priority(j)
  if good and score >= bestScore:
    bestScore, bestG = score, Gp
```

**Comment by Tyilo** [comment]
Merged in 5e0d46cf39365a7845f72ed251103eb8c3c72257.
